File: src/main/java/com/heekng/celloct/entity/Manager.java
package com.heekng.celloct.entity;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.*;

import static javax.persistence.FetchType.*;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Manager extends BaseTimeEntity{

    @Id @GeneratedValue
    @Column(name = "manager_id")
    private Long id;

    private String name;

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "shop_id")
    private Shop shop;

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "member_id")
    private Member member;

    @Builder
    public Manager(Shop shop, Member member, String name) {
        this.shop = shop;
        this.member = member;
        this.name = name;
        shop.getManagers().add(this);
        member.getManagers().add(this);
    }

    public void updateInfo(String name) {
        this.name = name;
    }

    public void addShop(Shop shop) {
        this.shop = shop;
        shop.getManagers().add(this);
    }

    public void addMember(Member member) {
        this.member = member;
        member.getManagers().add(this);
    }
}

File: src/main/kotlin/com/heekng/celloct/entity/Manager.kt
package com.heekng.celloct.entity

import javax.persistence.*

@Entity
@Table(
    uniqueConstraints = [UniqueConstraint(
        name = "manager_member_id_shop_id_unique",
        columnNames = ["member_id", "shop_id"]
    )]
)
class Manager(
    @Id @GeneratedValue @Column(name = "manager_id")
    val id: Long? = null,

    @Column(name = "name", nullable = false)
    var name: String,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "shop_id", nullable = false)
    var shop: Shop,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", nullable = false)
    var member: Member,
) : BaseTimeEntity() {

    companion object {
        fun fixture(
            shop: Shop,
            member: Member,
            name: String,
        ): Manager {
            val manager = Manager(
                shop = shop,
                member = member,
                name = name,
            )
            shop.managers.add(manager)
            member.managers.add(manager)
            return manager
        }
    }

    fun updateInfo(name: String) {
        this.name = name
    }

    fun addShop(shop: Shop) {
        this.shop = shop
        shop.managers.add(this)
    }

    fun addMember(member: Member) {
        this.member = member
        member.managers.add(this)
    }
}